src/WCSimPrimaryGeneratorAction.cc

```diff
@@ -905,8 +907,92 @@ void WCSimPrimaryGeneratorAction::GeneratePrimaries(G4Event* anEvent)
       	*/
       }
 
+    } else if (useDataTableEvt) {
+    // Setup local variables to store the data table values
+    G4int index;
+    G4int pdgid;
+    G4double ene;
+    G4ThreeVector pos;
+    G4ThreeVector dir;
+    G4double t;
+
+    // Check if the input file is open
+    if (!inputFile.is_open()) {
+      G4cout << "Error: Input file is not open" << G4endl;
+      G4cout << "Set a vector file using the command /mygen/vecfile <FILENAME>"
+             << G4endl;
+      exit(-1);
     }
-  else if(useCosmics){
+
+    // Flag for first event
+    G4bool firstParticle = true;
+
+    // Line position before the next particle
+    std::streampos lastLinePos;
+
+    // Counter for the number of particles per event
+    G4int nParticles = 0;
+
+    // Read the data table
+    std::string line;
+
+    // Skip any lines that start with a #
+    while (std::getline(inputFile, line)) {
+      if (line.empty() || line[0] == '#') {
+        continue;
+      }
+
+      // Buffer to convert between string and other variables
+      std::istringstream buffer(line);
+
+      // Load information into local variables
+      buffer >> index >> pdgid >> ene >> pos[0] >> pos[1] >> pos[2] >> dir[0] >>
+          dir[1] >> dir[2] >> t;
+
+      if (index == 0 && firstParticle == false) {
```

Comment:
```suggestion
      // We've finished reading all the particles constituing the event n we're generating
      // and we've also read the first line of event n+1
      // So rewind to the start of event n+1, for when we call this function again for the next event
      if (index == 0 && firstParticle == false) {
```

Just checking I understand what you're doing. And I think putting a comment in to explain it is a good idea. So please modify the comment for accuracy!

Reply:
Hi Tom! Thanks for the feedback and help.

Yeah, you were right. I was missing a couple of commits that actually contained the definitions in the messenger class and a comment explaining the rewind behaviour in the generator.

I've added everything else: check for if the generator reaches the end of the datatable before `BeamOn` is reached, check for the maximum number of vertices etc.

I think that I've also sorted out setting the information that's written to the output file, though I may have missed some. I'd appreciate it if you could double-check that if you get the chance, please.

Also, sorry the commits are such a mess; I've got an auto-formatter installed that I hadn't realised had made **many** _**many**_ changes to a lot of files in the same commits as actual work.

The pull from `WCSim/WCSim/develop` has also been done and _should_ be sorted now.

Let me know if there are any other issues!